### PR TITLE
Make `BasicClassifier` accept any metric

### PR DESCRIPTION
**Pull request by bryant1410**
Make `BasicClassifier` accept any metric, passed as args.

I didn't add tests yet because I want to receive feedback first. I'm leaving some comments on the code.

### CODE REVIEW

File: allennlp/training/metrics/metric.py [hunk 1]
```diff
@@ -11,7 +12,10 @@ class Metric(Registrable):
     """
 
     def __call__(
-        self, predictions: torch.Tensor, gold_labels: torch.Tensor, mask: Optional[torch.Tensor]
+        self,
+        predictions: torch.Tensor,
+        gold_labels: torch.Tensor,
+        mask: Optional[torch.Tensor] = None,
```
Comment (bryant1410) [inline allennlp/training/metrics/metric.py]: I changed this because mypy was complaining about the line that says `metric(logits, label)`, stating there's a missing param, which is actually true. I could have set it to `None`, but then saw the documentation and it says "default = None", which is actually false, so I changed it.

File: allennlp/training/metrics/metric.py [hunk 2]
```diff
@@ -34,6 +38,26 @@ def get_metric(
         """
         raise NotImplementedError
 
+    def get_metric_name_value_pairs(
```
Comment (bryant1410) [inline allennlp/training/metrics/metric.py]: This function is necessary to have a homogeneous output out of `get_metric`. It can be here, or it could be a util function maybe.

File: allennlp/training/metrics/metric.py [hunk 3]
```diff
@@ -34,6 +38,26 @@ def get_metric(
...
+        value = self.get_metric(reset)
+        if isinstance(value, collections.abc.Mapping):
+            for sub_name, sub_value in value.items():
+                if isinstance(sub_value, collections.abc.Iterable):
+                    for i, sub_value_i in enumerate(sub_value):
+                        yield f"{sub_name}_{i}", sub_value_i
+                else:
+                    yield sub_name, sub_value
+        elif isinstance(value, collections.abc.Iterable):
+            for i, sub_value in enumerate(value):  # type: ignore
```
Comment (bryant1410) [inline allennlp/training/metrics/metric.py]: mypy associates the previous variable (`sub_value`) with the same name the same type as this one, so I added this ignore.

File: allennlp/training/metrics/metric.py [hunk 4]
```diff
@@ -34,6 +38,26 @@ def get_metric(
...
+                yield f"{default_name}_{i}", sub_value  # type: ignore
```
Comment (bryant1410) [inline allennlp/training/metrics/metric.py]: Even when changing the variable name, mypy says `sub_value` is of type `object` instead of `float`.

File: allennlp/training/metrics/metric.py [hunk 5]
```diff
@@ -34,6 +38,26 @@ def get_metric(
         """
         raise NotImplementedError
 
+    def get_metric_name_value_pairs(
+        self, default_name: str, reset: bool = False
```
Comment (bryant1410) [inline allennlp/training/metrics/metric.py]: Maybe a default default name can be created from the underlying class name :man_shrugging:

File: allennlp/training/metrics/metric.py (hunk 2 above)
Comment (bryant1410) [reply allennlp/training/metrics/metric.py]: Another option is to change `get_metric`.

**Comment by matt-gardner** [comment]
I think the changes to the basic classifier are fine, but I don't really see the value in the changes to `Metric`.  You're trying to make a uniform API for `Metric`, but it doesn't actually exist, unless you do something like `*args`.  Mypy is just going to complain.  We could try to make a uniform API by adding subclasses like `ClassificationMetric`, but that's just confusing and not worth the complexity.

**Comment by bryant1410** [comment]
> You're trying to make a uniform API for Metric, but it doesn't actually exist, unless you do something like *args. Mypy is just going to complain.

Do you mean when calling `metric(logits, label)`?

**Comment by bryant1410** [comment]
> You're trying to make a uniform API for Metric, but it doesn't actually exist, unless you do something like *args. Mypy is just going to complain.

Do you mean when calling `metric(logits, label)`?

**Comment by bryant1410** [comment]
Mypy doesn't after adding the `= None` to `mask` because `metric(logits, label)` meets the contract, while some subclasses break it.

What if I do `metric(predictions=..., gold_labels=...)` in a try-except, and say to the user that it supports metrics that only support that in case of exception?

An option that goes further (non-exclusive with the previous one): change `__call__` to receives `**kwargs` and state in the docstrings the typical behavior, asking the subclasses to follow a common naming (`predicitons` and `gold_labels`)?

**Comment by DeNeutoy** [comment]
I'm less sure whether these changes are fine - you cannot pass any metric, so the type annotation as it stands is not correct. Why do you need to parametrise this? This is not a pattern we want people using our models to replicate, because they will get confused about which metrics can be passed to which models. I don't think we should accept this change, the Metrics are actually a great part of the library already, and their usage has never raised any questions.

**Comment by DeNeutoy** [comment]
> What if I do `metric(predictions=..., gold_labels=...)` in a try-except, and say to the user that it supports metrics that only support that in case of exception?
> 
Instead, you should just have fixed metrics in your model

> An option that goes further (non-exclusive with the previous one): change `__call__` to receive `**kwargs` and state in the docstrings the typical behavior, asking the subclasses to follow a common naming (`predicitons` and `gold_labels`)?

But in the case you pass the wrong arguments in kwargs, you will definitely be also passing the wrong arguments as args.

**Comment by matt-gardner** [comment]
@DeNeutoy, additional context [here](https://github.com/allenai/allennlp/issues/3824#issuecomment-589710136) (in the whole issue, but particularly that comment).  And yes, I agree with Mark that the model changes are too general.  I think it's fine to make it accept more metrics, but if we are going to do that, it should be more targeted than this.  As I said at the bottom of the referenced comment, the classification model could "have a list of allowable metrics, and a configuration option that allows a user to select them".  Allowing any metric like this is just asking for errors.

**Comment by matt-gardner** [comment]
And in general, if you want to make your model general enough to allow taking arbitrary metrics, it's really easy to do - just change the constructor arguments, as you do here.  The question is what our example models in the library should do.  And the way they are is probably good enough, though I could be convinced to allow, e.g., F1 or other things as options.

**Comment by matt-gardner** [comment]
But even then, we really want to be encouraging people to make their own models, not just use the ones we have out of the box.  Copying our model code and modifying it is just fine, and a major intended usage.  And so, given that, just having a comment in the original model saying that you can change this to F1 or something else if that's what you want in your model is maybe even good enough by itself.

**Comment by bryant1410** [comment]
I still think the underlying problem is that models are coupled with metrics, as discussed in #3824. They have certain inputs (e.g., `tokens` and `label`) and outputs (e.g., `logits`, `probs` and `loss`), but they also have metrics, instead of the metrics being computed from their outputs. But, anyway.

Okay, so I change it to instead just accept F1 as well as Accuracy? With options for the beta value and the type of average?

**Comment by matt-gardner** [comment]
So, with talking about beta and average parameters, it's starting to get very messy, which makes me agree even more with @DeNeutoy that we don't need this change.  There's no reason for the example model we have in here to be a super-flexible, I-can-do-everything classifier.  That's probably not what any user would want to be writing, and having that be our example would scare people off as too confusing.

**Comment by bryant1410** [comment]
So maybe just a comment saying to copy and paste if you want flexibility?

**Comment by matt-gardner** [comment]
If we go that route, then we should basically be annotating all kinds of things in that example code.  That might be better suited to a template project, and leave the fact that you can modify stuff implicit in the stuff in here.
